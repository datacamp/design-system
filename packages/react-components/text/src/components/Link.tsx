import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css } from '@emotion/core';
import React, { ReactElement } from 'react';

import baseStyle from '../baseStyle';

interface LinkProps {
  /**
   * The content of the Link.
   */
  children: string;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * The url to navigate to, when the user clicks.
   */
  href: string;
}

const linkStyle = css(baseStyle, {
  color: tokens.color.opaque.primary.value.hex,
  cursor: 'pointer',
  fontSize: tokens.size.font[300].value,
  fontWeight: tokens.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
  textDecoration: 'underline',
});

const InternalLink = ({
  className,
  dataAttributes,
  children,
  innerRef,
  href,
}: LinkProps & { innerRef?: React.Ref<HTMLAnchorElement> }): ReactElement => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  return (
    <a
      className={className}
      css={linkStyle}
      href={href}
      ref={innerRef}
      {...parsedDataAttributes}
    >
      {children}
    </a>
  );
};

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => (
  <InternalLink {...props} innerRef={ref} />
));

export default Link;
