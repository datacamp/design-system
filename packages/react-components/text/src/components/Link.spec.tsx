import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import Link from './Link';

const testText = 'Some example text.';
const testClassName = 'example-class';
const testDataAttributes = { cy: 'othertest', example: 'testdata' };

describe('<Text />', () => {
  it('renders with a string', async () => {
    const { container } = await axeRender(
      <Link
        className={testClassName}
        dataAttributes={testDataAttributes}
        href="https://datacamp.com"
      >
        {testText}
      </Link>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('sets the text in the a', async () => {
    const { getByText } = await axeRender(
      <Link href="https://datacamp.com">{testText}</Link>
    );
    expect(getByText(testText)).toBeDefined();
  });

  it('sets the className', async () => {
    const { getByText } = await axeRender(
      <Link className={testClassName} href="https://datacamp.com">
        {testText}
      </Link>
    );
    expect(getByText(testText)).toHaveClass(testClassName);
  });

  it('sets the data attributes', async () => {
    const { getByText } = await axeRender(
      <Link dataAttributes={testDataAttributes} href="https://datacamp.com">
        {testText}
      </Link>
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-cy',
      testDataAttributes.cy
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-example',
      testDataAttributes.example
    );
  });
});
