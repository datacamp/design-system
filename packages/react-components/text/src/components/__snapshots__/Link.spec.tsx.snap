// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> renders with a string 1`] = `
.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans;
  font-style: normal;
  color: #007bb6;
  cursor: pointer;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  class="example-class emotion-0"
  data-example="testdata"
  data-item="othertest"
  href="https://datacamp.com"
>
  Some example text.
</a>
`;
